server/src/test/java/org/apache/druid/server/http/MetadataResourceTest.java

```diff
@@ -0,0 +1,182 @@
+import org.apache.druid.timeline.DataSegment;
+import org.apache.druid.timeline.SegmentStatusInCluster;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.core.Response;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+
+public class MetadataResourceTest
+{
+  private static final String DATASOURCE1 = "datasource1";
+  private static final String DATASOURCE2 = "datasource2";
+
+  private MetadataResource metadataResource;
+
+  private SegmentsMetadataManager segmentsMetadataManager;
+  private DruidCoordinator coordinator;
+  private HttpServletRequest request;
+
+  private final DataSegment dataSegment1 = new DataSegment(
+      DATASOURCE1,
+      Intervals.of("2010-01-01/P1D"),
+      "v0",
+      null,
+      null,
+      null,
+      null,
+      0x9,
+      10
+  );
+
+  private final DataSegment dataSegment2 = new DataSegment(
+      DATASOURCE1,
+      Intervals.of("2010-01-22/P1D"),
+      "v0",
+      null,
+      null,
+      null,
+      null,
+      0x9,
+      20
+  );
+
+  private final DataSegment dataSegment3 = new DataSegment(
+      DATASOURCE2,
+      Intervals.of("2010-01-01/P1M"),
+      "v0",
+      null,
+      null,
+      null,
+      null,
+      0x9,
+      30
+  );
+
+  private final DataSegment dataSegment4 = new DataSegment(
+      DATASOURCE2,
+      Intervals.of("2010-01-02/P1D"),
+      "v0",
+      null,
+      null,
+      null,
+      null,
+      0x9,
+      35
+  );
+
+  @Before
+  public void setUp()
+  {
+    // Create mock request
+    request = mock(HttpServletRequest.class);
+    doReturn(mock(AuthenticationResult.class)).when(request).getAttribute(AuthConfig.DRUID_AUTHENTICATION_RESULT);
+
+    // Mock SegmentsMetadataManager
+    segmentsMetadataManager = mock(SegmentsMetadataManager.class);
+    ImmutableDruidDataSource druidDataSource1 = new ImmutableDruidDataSource(
+        DATASOURCE1,
+        ImmutableMap.of(),
+        ImmutableList.of(
+            dataSegment1,
+            dataSegment2
+        )
+    );
+
+    ImmutableDruidDataSource druidDataSource2 = new ImmutableDruidDataSource(
+        DATASOURCE1,
+        ImmutableMap.of(),
+        ImmutableList.of(
+            dataSegment3,
+            dataSegment4
+        )
+    );
+
+    // Mock segments from cache and coordinator
+    DataSourcesSnapshot dataSourcesSnapshot = mock(DataSourcesSnapshot.class);
+    doReturn(dataSourcesSnapshot).when(segmentsMetadataManager).getSnapshotOfDataSourcesWithAllUsedSegments();
+
+    doReturn(ImmutableList.of(druidDataSource1, druidDataSource2)).when(dataSourcesSnapshot).getDataSourcesWithAllUsedSegments();
+    // Segment 4 is overshadowed
+
+    // Mock Coordinator
+    coordinator = mock(DruidCoordinator.class);
+    // Segment 1: Replication factor 2, not overshadowed
```

Comment:
General suggestion: Avoid such comments which don't provide any new info or explanation. Instead, prefer making the code itself a little more readable. Here, the code already seems readable enough to not need these comments.

Pretty much all the comments in this method can be removed since they don't provide much insight into the code.

Reply:
Thanks! This is pretty useful as a guideline.